test/helpers/policygen/models.go

```diff
@@ -0,0 +1,664 @@
+		err = t.l4.SetTemplate(&l4egress, t)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	switch kind := t.l7.kind; kind {
+	case ingress:
+		err = t.l7.SetTemplate(&l4ingress, t)
+	case egress:
+		err = t.l7.SetTemplate(&l4egress, t)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	if len(l4ingress) > 0 {
+		ingressMap[toPorts] = []rule{l4ingress}
+	}
+
+	if len(l4egress) > 0 {
+		egressMap[toPorts] = []rule{l4egress}
+	}
+
+	if len(ingressMap) > 0 {
+		specs = append(specs, Spec{
+			EndpointSelector: endpointSelector{
+				Matchlabels: map[string]string{
+					"id": t.DestPod},
+			},
+			Ingress: []rule{ingressMap},
+		})
+	}
+
+	if len(egressMap) > 0 {
+		specs = append(specs, Spec{
+			EndpointSelector: endpointSelector{
+				Matchlabels: map[string]string{
+					"id": t.SrcPod},
+			},
+			Egress: []rule{egressMap},
+		})
+	}
+
+	if len(specs) == 0 {
+		return "", nil
+	}
+
+	jsonOutput, err := json.Marshal(specs)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf(string(metadata), t.Prefix, jsonOutput, t), nil
+}
+
+// NetworkPolicyName returns the name of the NetworkPolicy
+func (t *TestSpec) NetworkPolicyName() string {
+	return fmt.Sprintf("%s_policy.json", t.Prefix)
+}
+
+// NetworkPolicyApply applies the cilium network policy in kubernetes and wait
+// until all  pods updates their status. It'll return an error if pods did not
+// update the state or policy cannot be applied
+func (t *TestSpec) NetworkPolicyApply() error {
+	policy, err := t.NetworkPolicyCreate()
+	if err != nil {
+		return fmt.Errorf("Network policy cannot be created prefix=%s: %s", t.Prefix, err)
+	}
+
+	if policy == "" {
+		//This only happens on L3:No Policy L4:No Policy L7:No Policy
+		log.Info("No policy so do not import it")
+		return nil
+	}
+
+	err = ioutil.WriteFile(t.NetworkPolicyName(), []byte(policy), os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("Network policy cannot be written prefix=%s: %s", t.Prefix, err)
+	}
+
+	_, err = t.Kub.CiliumImportPolicy(
+		helpers.KubeSystemNamespace,
+		fmt.Sprintf("%s/%s", helpers.BasePath, t.NetworkPolicyName()),
+		helpers.HelperTimeout)
+
+	if err != nil {
+		return fmt.Errorf("Network policy cannot be imported prefix=%s: %s", t.Prefix, err)
+	}
+	return nil
+}
+
+type connTestResultType struct {
+	kind   string
+	result ResultType
+}
+
+// GetConnectivyTest returns and array with the expected results of the given
+// connectivy test kind
+func (t *TestSpec) GetConnectivyTest(kind string) []connTestResultType {
+	return []connTestResultType{
+		connTestResultType{t.l3.kind, t.l3.tests.GetField(kind)},
+		connTestResultType{t.l4.kind, t.l4.tests.GetField(kind)},
+		connTestResultType{t.l7.kind, t.l7.tests.GetField(kind)}}
+}
+
+// GetTestExpects returns a map with the connTestType and the exepected result
+// based on the `testExpect`
+func (t *TestSpec) GetTestExpects() map[string]ResultType {
+	DetermineStatus := func(testType string) ResultType {
+		connTest := t.GetConnectivyTest(testType)
+
```

Comment:
I don't get why do we need these two for loops. As I understand, `DetermineStatus` returns what result is expected from give testType? What does it matter which rules are matched first?

Reply:
Easy, this is an example rule:

IngressRule on pod dest:
-- allow PORT 8080 --> ResultTimeout

EgressRule on src pod:
-- only access to /private/ url to port 80:

Testcase:

- If src pod made an http request to  http://destPod:80/private --> ResultTimeout Correct. (The rule that match is  the IngressRule)
If src pod made an http request to  http://destPod:80/public --> ResultAuth --> The expected was ResultTimeout

So, we need to check the egress  before the ingress part to determine the status of the expected result